docs/daemon_design.md

```diff
@@ -0,0 +1,685 @@
+
+```
+RollbackDeploymentResponse RollbackDeployment(RollbackDeploymentRequest)
+
+RollbackDeploymentRequest {
+    EnvironmentName: string
+    EnvironmentVersion: uuid
+}
+```
+
+### Stopping a deployment
+Stopping a deployment will halt the in-progress deployment if one exists. The started tasks will remain untouched but the deployment will not continue. The environment will be set to inactive to prevent tasks from being started on new instances joining the cluster.
+
+```
+StopDeploymentResponse StopDeployment(StopDeploymentRequest)
+
+StopDeploymentRequest {
+    EnvironmentName: string
+    DeploymentID: string
+}
+```
+
+### Deleting an environment
+An environment cannot be deleted if it has an in-progress deployment started by a user (if there are in-progress deployments started by new instance monitors, those will be stopped). The in-progress deployment needs to be stopped before the environment can be deleted.
+
+```
+DeleteEnvironmentResponse DeleteEnvironment(DeleteEnvironmentRequest)
+
+DeleteEnvironmentRequest {
+   Name: string
+}
+```
+### Getting deployment and environment state
+```
+GetEnvironmentResponse GetEnvironment(GetEnvironmentRequest)
+
+GetEnvironmentRequest {
+    EnvironmentName: string
+}
+
+GetEnvironmentResponse {
+    Environment object without full deployment history. To get deployment history call ListDeployments for the environment.
+}
+
+ListEnvironmentsResponse ListEnvironments(ListEnvironmentsRequest)
+
+ListEnvironmentsRequest {
+    EnvironmentType: [Daemon, Service, etc] (optional)
+}
+
+ListEnvironmentsResponse {
+    List of environments {
+        EnvironmentName: string
+        EnvironmentState: [active, inactive]
+        EnvironmentHealth: [healthy, unhealthy]
+        ActiveDeployment: Deployment if there is a pending or in-progress one
+        EnvironmentType: [Daemon, Service, etc]
+    }
+}
+
+GetDeploymentResponse GetDeployment(GetDeploymentRequest)
+
+GetDeploymentRequest {
+    EnvironmentName: string
+    DeploymentID: string
+}
+
+GetDeploymentResponse {
+    Deployment object
+}
+
+GetDeploymentsByState? or should this be an optional field in getDeployment if ID is not passed
+
+ListDeploymentsResponse ListDeployments(ListDeploymentsRequest)
+
+ListDeploymentsRequest {
+    EnvironmentName: string
+}
+
+ListDeploymentsResponse {
+    list of deployments reverse-time sorted (latest first) {
+        EnvironmentName: string
+        DeploymentID: string
+        DeploymentState: {in-progress, x/n complete etc}
+    }
+}
+```
+
+## Design
+### Architecture
+![Architectural Diagram](images/diagram.png)
+
+### Service Descriptions
+
+#### Blox Frontend
+Responsible for authentication, authorization, throttling, input validation, user facing API. It interacts with the data service to create environment and deployment objects in the database.
```

Comment:
@jhspaybar These will be new APIs but they will be available as part of the AWS SDK. If you were hosting this "locally", i.e. in your account, you will be able to set up the APIs in your API gateway but I'm not sure if you'll be able to override the endpoints in the AWS SDK to point to your API gateway. You might be able to set the endpoint on the sdk to point to a non-aws service but I'm not sure what validation they do and if that will work. Let me get back to you on that.